test/kubernetes/integration_test.go

```diff
@@ -674,6 +806,85 @@ func waitForPod(client kubernetes.Interface, name string) error {
 	return err
 }
 
+// waitForVolumeCapacityChange waits for the given pod name to be running
+func waitForVolumeCapacityChange(client kubernetes.Interface, name string, resourceList v1.ResourceList) error {
+	var err error
+	stopCh := make(chan struct{})
+
+	go func() {
+		select {
+		case <-time.After(time.Minute * 5):
+			err = errors.New("timing out waiting for pv state")
+			close(stopCh)
+		case <-stopCh:
+		}
+	}()
+
+	watchlist := cache.NewListWatchFromClient(client.CoreV1().RESTClient(),
+		"persistentvolumes", v1.NamespaceAll, fields.Everything())
+	_, controller := cache.NewInformer(watchlist, &v1.PersistentVolume{}, time.Second*1,
+		cache.ResourceEventHandlerFuncs{
+			UpdateFunc: func(o, n interface{}) {
+				volume := n.(*v1.PersistentVolume)
+				if name != volume.Name {
+					return
+				}
+				if volume.Status.Phase == v1.VolumeFailed {
+					err = errors.New("Persistent volume status is Failed")
+					close(stopCh)
+					return
+				}
+
+				if volume.Status.Phase == v1.VolumeBound && volume.Spec.Capacity["storage"] != resourceList["storage"] {
+					close(stopCh)
```

Comment:
If we also store the updated object here and return it after the controller has stopped, we can save the call to the API to fetch the PV resource again at line 355.

Similar for PVC.